proposals/per-extension-language-preferences.md

```diff
@@ -0,0 +1,234 @@
+
+### Schema
+
+```ts
+/**
+ * Get the extension displayed language.
+ * return the language tag, e.g. en-US.
+ */
+i18n.getCurrentLanguage() : string
+
+/**
+ * Set the user preferred language.
+ * if code is null, revert to the unset state.
+ * if code is not valid, return a rejected Promise.
+ * else return a Promise resolved when the operation is complete.
+ */
+i18n.setCurrentLanguage(code: string) : Promise<void>
+
+/**
+ * Get all languages that supported by this extension.
+ * return a Promise resolved with an array of language tags.
+ */
+i18n.getAllLanguages(): Promise<string[]>
+
+/**
+ * After changing to a new language, the browser triggers a language changed event.
+ * The callback is function(code: string) : void.
+ * The code parameter in the callback is the new language.
+ */
+i18n.onLanguageChanged.addListener(callback)
+```
+
+### Behavior
+
+#### Behavior of `i18n.getCurrentLanguage()`
+
+This method return the language that the extension is displayed in.
+
+- If the extension doesn't use `browser.i18n` (there is no "_locales" directory), return `undefined`.
+- If the preferred language is not set by `i18n.setCurrentLanguage()`, returns the current language used by `i18n.getMessage()`, assuming that all languages support all possible keys.
+- If the preferred language is set by `i18n.setCurrentLanguage()`, and the extension supports this language, then return this language.
+- If the preferred language is set by `i18n.setCurrentLanguage()`, but the extension doesn't support this language (no message file for this language), then treat as if the preferred language is not set. This is an edge case, for example, the language was removed when the extension was upgraded.
+
+##### Use Case 1: use the extension displayed language with other locale-related APIs.
+
+For example, if the preferred language is not set, the language of the browser UI is English (en-US), the extension supports French (default) and Japanese, then `i18n.getCurrentLanguage()` return `"fr"` because `i18n.getMessage()` return French message at this situation.
+
+```js
+// use current extension displayed language to format date
+const locale = browser.i18n.getCurrentLanguage();
+const today_date = new Date().toLocaleString(locale); // format date string in 'fr'
+
+// if don't use current extension displayed language, it may be formatted in other locale
+const today_date = new Date().toLocaleString(); // format date string in 'en-US'
+```
+
+##### Use Case 2: show the current language in the extension language selection menu.
+
+#### Behavior of `i18n.setCurrentLanguage(code)`
+
+This method sets the user preferred language for this extension.
+
+- If the extension doesn't use `browser.i18n` (there is no "_locales" directory), return a rejected Promise.
```

Comment:
Do we need this line? I think the line below covers it and I think that aligns more closely with how this would be implemented in practice.

Reply:
From an implementation perspective, this is a bit redundant. But, from a specification perspective, this situation is slightly different from the following situation.

Reply:
I think I'd still be in favour of removing this. It doesn't change anything about the behaviour and having the spec guide implementors seems like a desirable thing.

Reply:
I removed this line.